app/src/main/java/com/duckduckgo/app/privacy/ui/WhitelistActivity.kt

```diff
@@ -0,0 +1,236 @@
+            setMessage(getString(R.string.whitelistEntryDeleteConfirmMessage, entry.domain).html(this.context))
+            setPositiveButton(android.R.string.yes) { _, _ -> viewModel.onEntryDeleted(entry) }
+            setNegativeButton(android.R.string.no) { _, _ -> }
+        }.create()
+
+        dialog?.dismiss()
+        dialog = deleteDialog
+        deleteDialog.show()
+    }
+
+    private fun showWhitelistFormatError() {
+        Toast.makeText(this, R.string.whitelistFormatError, Toast.LENGTH_LONG).show()
+    }
+
+    override fun onDestroy() {
+        dialog?.dismiss()
+        super.onDestroy()
+    }
+
+    companion object {
+        fun intent(context: Context): Intent {
+            return Intent(context, WhitelistActivity::class.java)
+        }
+    }
+
+    class WhitelistAdapter(private val viewModel: WhitelistViewModel) : Adapter<WhitelistViewHolder>() {
+
+        var entries: List<UserWhitelistedDomain> = emptyList()
+            set(value) {
+                field = value
+                notifyDataSetChanged()
+            }
+
+        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WhitelistViewHolder {
+            val inflater = LayoutInflater.from(parent.context)
+            val view = inflater.inflate(R.layout.view_whitelist_entry, parent, false)
+            return WhitelistViewHolder(view, view.domain, view.overflowMenu)
+        }
+
+        override fun onBindViewHolder(holder: WhitelistViewHolder, position: Int) {
+            val entry = entries[position]
+            holder.domain.text = entry.domain
+            holder.root.setOnClickListener {
+                viewModel.onEditRequested(entry)
+            }
+
+            val overflowContentDescription = holder.overflowMenu.context.getString(R.string.whitelistEntryOverflowContentDescription, entry.domain)
+            holder.overflowMenu.contentDescription = overflowContentDescription
+            holder.overflowMenu.setOnClickListener {
+                showOverflowMenu(holder.overflowMenu, entry)
+            }
+        }
+
+        private fun showOverflowMenu(overflowMenu: ImageView, entry: UserWhitelistedDomain) {
+            val popup = PopupMenu(overflowMenu.context, overflowMenu)
+            popup.inflate(R.menu.whitelist_individual_overflow_menu)
+            popup.setOnMenuItemClickListener {
+                when (it.itemId) {
+                    R.id.edit -> {
+                        viewModel.onEditRequested(entry)
+                        true
+                    }
+                    R.id.delete -> {
+                        viewModel.onDeleteRequested(entry)
+                        true
+                    }
+                    else -> false
+                }
+            }
+            popup.show()
+        }
+
+        override fun getItemCount(): Int {
+            return entries.size
+        }
+    }
+
+    class WhitelistViewHolder(
+        val root: View,
```

Comment:
I could get rid of it but then I'd need to attach a more generic click listener and I think this is a little better as it eliminates some ugliness in that area. Resolving, since this is a nit, but happy to discuss it more if you have questions.